pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/PinotTablePartitionRule.java

```diff
@@ -119,12 +113,15 @@ public void run()
       //We define have a desirable segment size OPTIMAL_SIZE_PER_SEGMENT
       //Divide the size of data coming in on a given day by OPTIMAL_SIZE_PER_SEGMENT we get the number of partitions.
       if (!_input.getOverWrittenConfigs().getPartitionConfig().isNumPartitionsOfflineOverwritten()) {
-        _output.getPartitionConfig().setNumPartitionsOffline((int) (optimalOfflinePartitions));
+        int optimalOfflinePartitions = (int) _output.getSegmentSizeRecommendations().getNumSegments();
+        _output.getPartitionConfig().setNumPartitionsOffline(optimalOfflinePartitions);
       }
     }
     if (_input.getTableType().equalsIgnoreCase(HYBRID)) {
       if (!_input.getOverWrittenConfigs().getPartitionConfig().isNumPartitionsOfflineOverwritten()) {
-        _output.getPartitionConfig().setNumPartitionsOffline(Math.min(optimalOfflinePartitions, numKafkaPartitions));
+        int optimalOfflinePartitions =
+            Math.min((int) _output.getSegmentSizeRecommendations().getNumSegments(), numKafkaPartitions);
+        _output.getPartitionConfig().setNumPartitionsOffline(optimalOfflinePartitions);
```

Comment:
Do we need to add additional check on cardinality further below where we pick the column to partition on? I remember discussing the case where cardinality < numPartitions

Reply:
Done.